src/plugins/GLesmos/glesmosCanvas.ts

```diff
@@ -155,72 +104,163 @@ export function initGLesmosCanvas() {
     const p2m = transforms.pixelsToMath;
     c.width = w;
     c.height = h;
+
     gl.viewport(0, 0, c.width, c.height);
+
+    glesmosSDFrequiredSteps = Math.ceil(Math.log2(Math.max(w, h)));
+
     cornerOfGraph = [p2m.tx, p2m.sy * h + p2m.ty];
     sizeOfGraph = [p2m.sx * w, -p2m.sy * h];
+
+    for (const tex of textures) {
+      // resize the framebuffer textures
+      gl.bindTexture(gl.TEXTURE_2D, tex);
+      gl.texImage2D(
+        gl.TEXTURE_2D,
+        0,
+        gl.RGBA32F,
+        w,
+        h,
+        0,
+        gl.RGBA,
+        gl.FLOAT,
+        null
+      );
+    }
+
+    // resize the cache
+    gl.bindTexture(gl.TEXTURE_2D, cacheTexture);
+    gl.texImage2D(
+      gl.TEXTURE_2D,
+      0,
+      gl.RGBA32F,
+      w,
+      h,
+      0,
+      gl.RGBA,
+      gl.FLOAT,
+      null
+    );
   };
 
-  //= =========================== WEBGL STUFF ==========================
-  const fullscreenQuadBuffer = gl.createBuffer();
-  gl.bindBuffer(gl.ARRAY_BUFFER, fullscreenQuadBuffer);
-  gl.bufferData(gl.ARRAY_BUFFER, FULLSCREEN_QUAD, gl.STATIC_DRAW);
+  const setupGLesmosEnvironment = (program: GLesmosProgram) => {
+    // vertex buffer
+    gl.enableVertexAttribArray(program.vertexAttribPos);
+    gl.vertexAttribPointer(program.vertexAttribPos, 2, gl.FLOAT, false, 8, 0);
+    gl.bindBuffer(gl.ARRAY_BUFFER, texCoordBuffer);
+
+    // uniforms
+    gl.uniform2fv(program.size, sizeOfGraph);
+    gl.uniform2fv(program.corner, cornerOfGraph);
+    gl.uniform1f(program.NaN, NaN);
+    gl.uniform1f(program.Infinity, Infinity);
+  };
+
+  //= ================ WEBGL FUNCTIONS ================
+
+  const buildGLesmosShaders = (
+    id: string,
+    deps: string,
+    chunk: GLesmosShaderChunk
+  ) => {
+    glesmosCache = glesmosGetCacheShader(gl, id, chunk, deps);
+    glesmosFinalPass = glesmosGetFinalPassShader(gl, id, chunk);
 
-  let glesmosShaderProgram: WebGLProgram | undefined;
-
-  const setGLesmosShader = (shaderCode: string, id: string) => {
-    glesmosShaderProgram = getShaderProgram(gl, shaderCode, id, () => {
-      return {
-        fragmentSource: GLESMOS_FRAGMENT_SHADER.replace(
-          /\/\/REPLACE_WITH_GLESMOS[\s\S]*\/\/REPLACE_WITH_GLESMOS_END/g,
-          shaderCode
-        ),
-        vertexSource: VERTEX_SHADER,
-      };
-    });
+    if (chunk.line_width === 0) {
+      // TODO: this is bad, globals are bad
```

Comment:
Yeah this whole file has been over-globaled from the beginning. A whole bunch of closured variables with a bunch of functions. Perfect use-case for a class; a refactor would be reasonable after the PR is merged.

Reply:
This is part of the reason why I created the `glesmosShaderProgram` class :)